# Comparative Analysis of Breast Cancer Studies Replicated. 

## About this study:
This project was inspired by a research article  by Muhammet Fatih AK, [A Comparative Analysis of Breast Cancer Detection and Diagnosis Using Data Visualization and Machine Learning Applications](https://www.mdpi.com/2227-9032/8/2/111). 

## Introduction:
   It's a very well-known fact that cancer is the most complicated of all the disease types. Cancer and it's treatment, as seen from modern day research, is indiviual dependent. Meaning that camcer is as a disease behaves in various ways, and how the immunue system responds to treatment is different. This is why there is a new focus on personalized cancer therapies in research and medicine. A great tool for the developmet of these therapies is data science and ML techniques, but these tools are also great at predicting early cancer diagnosis. The research paper referenced in this artcle is an attempt to do a comparative analysis of breast cancer using dataset from different studies to compare their own Machine Learning Algorithm's outcomes using a data set from University of Wisconsin. 
   
   
## Dataset Used: 
   Since I couldn't find the original dataset used in this article. I went on to Google to find the original dataset by Dr. Wolberg. The dataset is describes the characteristics of the cancerous cell and the patient ids and diagnosis. The data can be foudn at [UC Irvin Machine Learning Repository](https://archive.ics.uci.edu/dataset/17/breast+cancer+wisconsin+diagnostic). We can also use use the university's remote server to downlaod the data: 
 [university's remote server](ftp ftp.cs.wisc.edu cd math-prog/cpo-dataset/machine-learn/WDBC/)
 and the command line: 
 ftp ftp.cs.wisc.edu
 
## Steps To use the FTP reote server: 
# Download: 

    1. Install FTP Client, if you are accessing the database outside the university. Some examples are: 
    + 1a. FileZilla
    + 1b. Cyberduck
    + 1c. Command-Line FTP
    2. Launch the FTP server on your laptop. 
  
# Steps: 
    1. Open the terminal on your desktop
    2. tye this on the command line: 
      ```ftp ftp.cs.wisc.edu```
    3. Enter the password and the username if needed 
    4. Once you are inside the server enter cd to change the directory, and use this code: 
    ```cd math-prog/cpo-dataset/machine-learn/WDBC/```
    5. Once inside use the commands ls or dir. 
      a. ls- is a linux command that list the files and directories in the current directory, associated with Unix-like systems 
      b. dir - is a linux command that lists the files and directories as well but its often associated with  Windows and MS-DOS                     environments r
    6. Now if we wanted to download the data we would use the get command like: 
    ```get example.txt```
    7. Finally if we want to quit the bash, we would use the command quit. 
  
  For the ease of this projet I will just download the database on my laptop.
  
## Downloading the Data: 

    1. First I downladed the dataset from the internet.
    2. Now we code so that we can read and present the data. 
    3. we imported this data on python. 


First we will need to install the packages for the library to access the data: 

In [7]:
pip install ucimlrepo


[notice] A new release of pip is available: 23.2.1 -> 23.3
[notice] To update, run: pip3 install --upgrade pip
Note: you may need to restart the kernel to use updated packages.


Our requirement was already satisfied. Next we will need to import that data. In order to import the data on our machine, we needed to import ssl, which is a secured socket layer. We have to disable this so we we can extract the data. The command below allows us to trust the SSL certficate to do so.  

In [8]:
import ssl

# Disable SSL certificate verification
ssl._create_default_https_context = ssl._create_unverified_context


In [9]:
from ucimlrepo import fetch_ucirepo
import pandas as pd

that we have imported our libraries and disabled the SSL certificates, now we can import our data. Here in this code we extracted our data directly from the UCI Machine Learning site. Since the data extracted was not ordered, we used the command .data.features to 

In [149]:
df = fetch_ucirepo(id=17) 
print(df.metadata) 
# variable information 
print(df.variables) 


{'uci_id': 17, 'name': 'Breast Cancer Wisconsin (Diagnostic)', 'repository_url': 'https://archive.ics.uci.edu/dataset/17/breast+cancer+wisconsin+diagnostic', 'data_url': 'https://archive.ics.uci.edu/static/public/17/data.csv', 'abstract': 'Diagnostic Wisconsin Breast Cancer Database.', 'area': 'Life Science', 'tasks': ['Classification'], 'characteristics': ['Multivariate'], 'num_instances': 569, 'num_features': 30, 'feature_types': ['Real'], 'demographics': [], 'target_col': ['Diagnosis'], 'index_col': ['ID'], 'has_missing_values': 'no', 'missing_values_symbol': None, 'year_of_dataset_creation': 1993, 'last_updated': 'Mon Jul 17 2023', 'dataset_doi': '10.24432/C5DW2B', 'creators': ['William Wolberg', 'Olvi Mangasarian', 'Nick Street', 'W. Street'], 'intro_paper': {'title': 'Nuclear feature extraction for breast tumor diagnosis', 'authors': 'W. Street, W. Wolberg, O. Mangasarian', 'published_in': 'Electronic imaging', 'year': 1993, 'url': 'https://www.semanticscholar.org/paper/53f0fbb42

In [171]:
print(df)


{'data': {'ids':            ID
0      842302
1      842517
2    84300903
3    84348301
4    84358402
..        ...
564    926424
565    926682
566    926954
567    927241
568     92751

[569 rows x 1 columns], 'features':      radius1  texture1  perimeter1   area1  smoothness1  compactness1  \
0      17.99     10.38      122.80  1001.0      0.11840       0.27760   
1      20.57     17.77      132.90  1326.0      0.08474       0.07864   
2      19.69     21.25      130.00  1203.0      0.10960       0.15990   
3      11.42     20.38       77.58   386.1      0.14250       0.28390   
4      20.29     14.34      135.10  1297.0      0.10030       0.13280   
..       ...       ...         ...     ...          ...           ...   
564    21.56     22.39      142.00  1479.0      0.11100       0.11590   
565    20.13     28.25      131.20  1261.0      0.09780       0.10340   
566    16.60     28.08      108.30   858.1      0.08455       0.10230   
567    20.60     29.33      140.10  1265.0      

## Comparing the outcomes using the two codes: 
So the code above this, shows all the features and the type of vraiables, however the problem I keep getting is the the diagnosis does not show up with the rest of the data. I personally would like to have the diagnosis in the same data.  So for now we will probably download the data, but we did successfully extract the data from the uciml website. 

In [172]:
breast_cancer = fetch_ucirepo(id=17) 
Features = breast_cancer.data.features
Features.head()


radius1  texture1  perimeter1   area1  smoothness1  compactness1  \
0    17.99     10.38      122.80  1001.0      0.11840       0.27760   
1    20.57     17.77      132.90  1326.0      0.08474       0.07864   
2    19.69     21.25      130.00  1203.0      0.10960       0.15990   
3    11.42     20.38       77.58   386.1      0.14250       0.28390   
4    20.29     14.34      135.10  1297.0      0.10030       0.13280   

   concavity1  concave_points1  symmetry1  fractal_dimension1  ...  radius3  \
0      0.3001          0.14710     0.2419             0.07871  ...    25.38   
1      0.0869          0.07017     0.1812             0.05667  ...    24.99   
2      0.1974          0.12790     0.2069             0.05999  ...    23.57   
3      0.2414          0.10520     0.2597             0.09744  ...    14.91   
4      0.1980          0.10430     0.1809             0.05883  ...    22.54   

   texture3  perimeter3   area3  smoothness3  compactness3  concavity3  \
0     17.33      184.60  2019.0       0.1622        0.6656      0.7119   
1     23.41      158.80  1956.0       0.1238        0.1866      0.2416   
2     25.53      152.50  1709.0       0.1444        0.4245      0.4504   
3     26.50       98.87   567.7       0.2098        0.8663      0.6869   
4     16.67      152.20  1575.0       0.1374        0.2050      0.4000   

   concave_points3  symmetry3  fractal_dimension3  
0           0.2654     0.4601             0.11890  
1           0.1860     0.2750             0.08902  
2           0.2430     0.3613             0.08758  
3           0.2575     0.6638             0.17300  
4           0.1625     0.2364             0.07678  

[5 rows x 30 columns]

We would also like to get a general summary of the data 

In [173]:
Features.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 569 entries, 0 to 568
Data columns (total 30 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   radius1             569 non-null    float64
 1   texture1            569 non-null    float64
 2   perimeter1          569 non-null    float64
 3   area1               569 non-null    float64
 4   smoothness1         569 non-null    float64
 5   compactness1        569 non-null    float64
 6   concavity1          569 non-null    float64
 7   concave_points1     569 non-null    float64
 8   symmetry1           569 non-null    float64
 9   fractal_dimension1  569 non-null    float64
 10  radius2             569 non-null    float64
 11  texture2            569 non-null    float64
 12  perimeter2          569 non-null    float64
 13  area2               569 non-null    float64
 14  smoothness2         569 non-null    float64
 15  compactness2        569 non-null    float64
 16  concavit

In [174]:
Features.describe()

radius1    texture1  perimeter1        area1  smoothness1  \
count  569.000000  569.000000  569.000000   569.000000   569.000000   
mean    14.127292   19.289649   91.969033   654.889104     0.096360   
std      3.524049    4.301036   24.298981   351.914129     0.014064   
min      6.981000    9.710000   43.790000   143.500000     0.052630   
25%     11.700000   16.170000   75.170000   420.300000     0.086370   
50%     13.370000   18.840000   86.240000   551.100000     0.095870   
75%     15.780000   21.800000  104.100000   782.700000     0.105300   
max     28.110000   39.280000  188.500000  2501.000000     0.163400   

       compactness1  concavity1  concave_points1   symmetry1  \
count    569.000000  569.000000       569.000000  569.000000   
mean       0.104341    0.088799         0.048919    0.181162   
std        0.052813    0.079720         0.038803    0.027414   
min        0.019380    0.000000         0.000000    0.106000   
25%        0.064920    0.029560         0.020310    0.161900   
50%        0.092630    0.061540         0.033500    0.179200   
75%        0.130400    0.130700         0.074000    0.195700   
max        0.345400    0.426800         0.201200    0.304000   

       fractal_dimension1  ...     radius3    texture3  perimeter3  \
count          569.000000  ...  569.000000  569.000000  569.000000   
mean             0.062798  ...   16.269190   25.677223  107.261213   
std              0.007060  ...    4.833242    6.146258   33.602542   
min              0.049960  ...    7.930000   12.020000   50.410000   
25%              0.057700  ...   13.010000   21.080000   84.110000   
50%              0.061540  ...   14.970000   25.410000   97.660000   
75%              0.066120  ...   18.790000   29.720000  125.400000   
max              0.097440  ...   36.040000   49.540000  251.200000   

             area3  smoothness3  compactness3  concavity3  concave_points3  \
count   569.000000   569.000000    569.000000  569.000000       569.000000   
mean    880.583128     0.132369      0.254265    0.272188         0.114606   
std     569.356993     0.022832      0.157336    0.208624         0.065732   
min     185.200000     0.071170      0.027290    0.000000         0.000000   
25%     515.300000     0.116600      0.147200    0.114500         0.064930   
50%     686.500000     0.131300      0.211900    0.226700         0.099930   
75%    1084.000000     0.146000      0.339100    0.382900         0.161400   
max    4254.000000     0.222600      1.058000    1.252000         0.291000   

        symmetry3  fractal_dimension3  
count  569.000000          569.000000  
mean     0.290076            0.083946  
std      0.061867            0.018061  
min      0.156500            0.055040  
25%      0.250400            0.071460  
50%      0.282200            0.080040  
75%      0.317900            0.092080  
max      0.663800            0.207500  

[8 rows x 30 columns]

In [175]:
### IMPORTING THE DATA USING A CSV FILE: 
data = pd.read_csv("/Users/kashishbhatt/Downloads/data.csv")
print(data)

           id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0      842302         M        17.99         10.38          122.80     1001.0   
1      842517         M        20.57         17.77          132.90     1326.0   
2    84300903         M        19.69         21.25          130.00     1203.0   
3    84348301         M        11.42         20.38           77.58      386.1   
4    84358402         M        20.29         14.34          135.10     1297.0   
..        ...       ...          ...           ...             ...        ...   
564    926424         M        21.56         22.39          142.00     1479.0   
565    926682         M        20.13         28.25          131.20     1261.0   
566    926954         M        16.60         28.08          108.30      858.1   
567    927241         M        20.60         29.33          140.10     1265.0   
568     92751         B         7.76         24.54           47.92      181.0   

     smoothness_mean  compa

In [177]:
## here I will obtain the number of diagnosis and the percentage of them: 
print("Number of Diagnosis:", data["diagnosis"].value_counts(), "\n", "_____________________")
print("Percentage of Diagnosis:", data["diagnosis"].value_counts() * 100 / len(data))


Number of Diagnosis: diagnosis
B    357
M    212
Name: count, dtype: int64 
 _____________________
Percentage of Diagnosis: diagnosis
B    62.741652
M    37.258348
Name: count, dtype: float64


## Logistic Regression:

Logistic regression is a Machine learning Technique, it is classified under supervised machine learning technique. Now one might ask, what is supervised technique? Well in machine learning there are two types of learning: supervised and un-supervised. Supervised machine learning is also known as classification data because it utilizes the use of labels in input and output data, while unsupervised learning doesn't but often is utilized to understand the relationships between inside the dataset. 

Logistic regression was first used in biological dataset in the 20th century and since then it has been widely used for social media based studies as well, as stated in the research paper used in this article. Its is best to use logistic regression when there is a binary classification. Meaning that there are only two outcomes like pass/fail, yes/no, diagnosed/not diagnosed, etc. The goal of logistic regression is to do predictive analysis, or in better terms estimate the probabilities of an event occurring. The outcome in this type of model is often coded as 0, which is the negative outcome, and 1, which is the positive outcome. So the dependent variable, that we are testing is bounded between 0 and 1. So that means that the predictive values should always fall within the range of 0 to 100%. The sigmoid function in the logistic regression assigns that probability. 

Mathematically, logistic regression calculates the probability or odd by this formula: 

$p = \frac{1}{1 + e^{-(b_0 + b_1)}}$

Logistic regression also has two phases, one is the forward propagation and backward propagation. So what are they? Forward propagation is when we multiply weights with features and since weights are often unknown we can assign random values to them. We finally then do a prediction based on threshold values, and once the prediction is done, the value of the prediction is then compared to the value of the values that were observed. and using that a loss function is generated. The loss function is what indicates how far the predicted value is from the real/observed value. backward propagation then comes in, whose goal is to update the weights according to the cost function. 

Here is the loss function: 

$Loss(y, \hat{y}) = -[y \cdot \log(\hat{y}) + (1 - y) \cdot \log(1 - \hat{y})]$

Here is the cost function: 

$J(\theta) = -\frac{1}{m} \sum_{i=1}^{m}[y^{(i)} \cdot \log(\hat{y}^{(i)}) + (1 - y^{(i)}) \cdot \log(1 - \hat{y}^{(i)})]$

The sigmoid function: 

$\sigma(x) = \frac{1}{1 + e^{-x}}$


### CODE FOR LOGISTIC REGRESSION: 


In [179]:
#importing the necessary libraries: 
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

In [181]:
from ucimlrepo import fetch_ucirepo 
  
# fetch dataset 
breast_cancer_wisconsin_diagnostic = fetch_ucirepo(id=17) 
  
# data (as pandas dataframes) 
X = breast_cancer_wisconsin_diagnostic.data.features 
y = breast_cancer_wisconsin_diagnostic.data.targets 

# so we have our data here according to machine learning techniques. 

# now we want to split the data: 
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# now we want to create a model and train it: 
model = LogisticRegression()
model.fit(X_train, y_train)

#And finally we want to predict the data's outcomes 

y_pred = model.predict(X_test)

# And now we want to evaluate the model. The model evaluation usually means that we are using common metrics to include the accuracy, precision, recall, F1-score and the confusion matrix.
accuracy = accuracy_score(y_test, y_pred)
report = classification_report(y_test, y_pred)
confusion = confusion_matrix(y_test, y_pred)

print(f'Accuracy: {accuracy}')
print(f'Classification Report:\n{report}')
print(f'Confusion Matrix:\n{confusion}')



Accuracy: 0.956140350877193
Classification Report:
              precision    recall  f1-score   support

           B       0.95      0.99      0.97        71
           M       0.97      0.91      0.94        43

    accuracy                           0.96       114
   macro avg       0.96      0.95      0.95       114
weighted avg       0.96      0.96      0.96       114

Confusion Matrix:
[[70  1]
 [ 4 39]]


/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/sklearn/utils/validation.py:1183: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/sklearn/linear_model/_logistic.py:460: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


### Analyzing the above details: 

Accuracy: 0.956140350877193

Classification Report:
|           | precision |   recall |   f1-score |   support |
|:----------|----------:|---------:|-----------:|----------:|
| B         |     0.95  |     0.99  |       0.97 |       71  |
| M         |     0.97  |     0.91  |       0.94 |       43  |
|           |           |          |            |           |
| accuracy  |     0.96  |     0.956 |       0.96 |      114  |
| macro avg |     0.96  |     0.95  |       0.95 |      114  |
| weighted avg |     0.96  |     0.96  |       0.96 |      114  |

Confusion Matrix:
|          |   |   |
|----------|--|--|
|    B      | 70 |  1 |
|    M     |  4 | 39 |


So this is a classification report with a confusion matrix. These provide us with the information regarding the logistic model for binary classification. In this report. the precision (which is the accuracy of positive predictions) shows 0.95 or 95% for begin and 0.97 or 97% for Malignant, the classification is B and M respectively. The recall is the true positive rate and its is the model's ability to identify all relevant instances. And here there is 0.99 or 99% positive recall for benign, meaning that 99% were accurately identified and for Malignant it was 0.91, or 91% were accurately identified. The F1 score is the mean of precision and recall, and its 0.97 for benign and 0.94 for Malignant. The support is the number of samples in each class. The over all accuracy of the model is 0.96 or 96% and 0.956 and 0.96 for benign and malignant respectively. Macro avg is the average of each column. 


The confusion matrix shows us the number of instances that were classified correctly and incorrectly. For Benign it was 70 samples that were correctly identified and 1 was incorrectly identified. For Malignant there were 39 samples that were correctly classified and 4 were incorrect. 
 

In [184]:
## CREATING A LOGISTIC MODEL DIAGRAM ACCORDING TO THE RESEARCH PAPER: 
import matplotlib.pyplot as plt 
from sklearn.linear_model import LogisticRegression
import numpy as np

# here we will selected the two features we want to visualize: 

feature1 = "radius_mean"
feature2 = "concavity_mean"

# we will then extract the two selected features from the data 
X_train_selected = X_train[:, [feature1, feature2]]
X_test_selected = X_test[:, [feature1, feature2]]


# now we will try to fit the logistic regression on the selected features 
model = LogisticRegression()
model.fit(X_train_selected, y_train)

# Now we will create a mesh grid to plot the decision boundary

x_min, x_max = X_train_selected[:, 0].min() - 1, X_train_selected[:, 0].max() + 1
y_min, y_max = X_train_selected[:, 1].min() - 1, X_train_selected[:, 1].max() + 1
xx, yy = np.meshgrid(np.arange(x_min, x_max, 0.01), np.arange(y_min, y_max, 0.01))
Z = model.predict(np.c_[xx.ravel(), yy.ravel()])
Z = Z.reshape(xx.shape)

# Plot the decision boundary and data points
plt.contourf(xx, yy, Z, alpha=0.4)
plt.scatter(X_train_selected[:, 0], X_train_selected[:, 1], c=y_train, marker='o')
plt.xlabel(f'Feature {feature1}')
plt.ylabel(f'Feature {feature2}')
plt.title('Logistic Regression Decision Boundary')
plt.show()

InvalidIndexError: (slice(None, None, None), ['radius_mean', 'concavity_mean'])

In [187]:
from sklearn.linear_model import LogisticRegression
import numpy as np
import matplotlib.pyplot as plt

# Define the selected features
feature1 = "radius_mean"
feature2 = "concavity_mean"

# Extract the two selected features from the data
X_train_selected = X_train[[feature1, feature2]]
X_test_selected = X_test[[feature1, feature2]]

# Fit the logistic regression on the selected features
model = LogisticRegression()
model.fit(X_train_selected, y_train)

# Create a mesh grid to plot the decision boundary
x_min, x_max = X_train_selected[feature1].min() - 1, X_train_selected[feature1].max() + 1
y_min, y_max = X_train_selected[feature2].min() - 1, X_train_selected[feature2].max() + 1
xx, yy = np.meshgrid(np.arange(x_min, x_max, 0.01), np.arange(y_min, y_max, 0.01))
Z = model.predict(np.c_[xx.ravel(), yy.ravel()])
Z = Z.reshape(xx.shape)

# Plot the decision boundary and data points
plt.contourf(xx, yy, Z, alpha=0.4)
plt.scatter(X_train_selected[feature1], X_train_selected[feature2], c=y_train, marker='o')
plt.xlabel(f'Feature: {feature1}')
plt.ylabel(f'Feature: {feature2}')
plt.title('Logistic Regression Decision Boundary')
plt.show()



KeyError: "None of [Index(['radius_mean', 'concavity_mean'], dtype='object')] are in the [columns]"

## K-Nearest Neighbor (KNN): 

K-Nearest Neighbor, also known as KNN, is a supervised ML techniques that uses proximity to make classifications. It is great to use in solving classification and regression problems. In this, the algorithms the data points that are the most similar to each other are often the closer to each other. In KNN, K is the numerical value for the K-nearest Neighbor. Its is one of those ML algorithms that don't have a training phase. The predictions in this model are made based on euclidean distance to the K-nearest neighbors. In the paper this algorithm was already applied to the the cancer dataset, using the labels malignant and benign. 

## Support Vector Machine (SVM)

Support Vector Machine (SVM) is one of the most popular ML techniques. This algorithm can be best used on small but complex datasets. Like logistic regression, the objective of SVM is to find the best hyperplane, however in SVM the goal is to use statistical approach instead of probabilistic approach. The N-dimension that classifies the data points in the hyper-plane diversifies based on the number of features. To get the best predictions you want to maximize the number of features. The downside to SVM is that there is a small trade off between large margin and the training data. Where the margin is the distance between hyper-plane and the nearest data points from each class. 

Here is the equation of the linear SVM: 
$y(x) = \text{sign}(\mathbf{w} \cdot \mathbf{x} + b)$

Where: 
y(x) = classification label for input vector
W = is the weight vector 
dot (.) = is the dot product
b = bias term 
sign(.) = is the sign function that returns: 
    +1 = positive value
    0 = neutral 
    -1 = negative value 


## Naive Bayes 



## Decision Trees  